src/invidious/helpers/utils.cr

```diff
@@ -471,3 +471,81 @@ def fetch_random_instance
 
   return filtered_instance_list.sample(1)[0]
 end
+
+struct LanguageEntry
+  class InvalidLanguageEntry < Exception
+  end
+
+  property tag, subtags, quality : Float64
+
+  def initialize(@tag : String, @subtags : Array(String), @quality : Float64)
+    @quality = @quality.clamp(0.0, 1.0)
+  end
+
+  def to_s(io)
+    io << tag
+    unless subtags.empty?
+      io << '-' << @subtags.join '-'
+    end
+  end
+
+  def self.from_string(language_str : String) : self
+    range_and_quality = language_str.split ';'
+
+    raise LanguageEntry::InvalidLanguageEntry.new if range_and_quality.empty?
+
+    if range_and_quality[1]?
+      quality = parse_quality(range_and_quality[1])
+    else
+      quality = 1.00
+    end
+
+    language_range = range_and_quality[0]
+    tags = language_range.split '-'
+
+    language_tag = tags[0]
+    subtags = tags[1..]
```

Comment:
Anything beyond the first sub-tag can likely be ignored; Invidious is unlikely to ever localize to anything that specific.

Reply:
Subtags can then be renamed to `territory`

Reply:
I can't recall what the MDN docs say on that, but shouldn't be the logic in `pt-BR;fr-FR,0.5;en,0.1` to check if `pt-BR` is available, then `pt | pt-*`, then `fr-FR`, then `fr | fr-*`, and finally `en | en-*`?

Reply:
That's what it currently does.

Reply:
Yeah the reason I just did an exact match is because my Accept-Language header was `en-CA,en-US;q=0.7,en;q=0.3`.
So if it was checked as `en-CA -> en -> en-*` it might skip `en-US` -- which is the next best locale quality-wise -- and choose `en` first instead. I guess a drawback would be if the browser doesn't specify a more general locale at the end of the specific ones then Invidious will just return the default of English like before.

I know this is a very unlikely scenario with the locales Invidious actually supports so maybe its not a big deal but it just seemed like confusing behaviour. But it's also possible I'm just misunderstanding.

If there is different matching behaviour desired I wouldn't mind implementing it.

Reply:
Subtags beyond the first are now ignored and I've changed the name to `region` as in the MDN page. Let me know if you would still prefer `territory` though.

Reply:
Ah, it's a bit more complex than what I expected:
https://datatracker.ietf.org/doc/html/rfc4647#section-2.3